subprojects/core/src/main/java/org/gradle/api/internal/project/LifecycleAwareProject.java

```diff
@@ -0,0 +1,1180 @@
+import org.gradle.internal.metaobject.BeanDynamicObject;
+import org.gradle.internal.metaobject.DynamicInvokeResult;
+import org.gradle.internal.metaobject.DynamicObject;
+import org.gradle.internal.model.ModelContainer;
+import org.gradle.internal.model.RuleBasedPluginListener;
+import org.gradle.internal.service.scopes.ServiceRegistryFactory;
+import org.gradle.invocation.IsolatedProjectEvaluationListenerProvider;
+import org.gradle.model.internal.registry.ModelRegistry;
+import org.gradle.normalization.InputNormalizationHandler;
+import org.gradle.normalization.internal.InputNormalizationHandlerInternal;
+import org.gradle.process.ExecResult;
+import org.gradle.process.ExecSpec;
+import org.gradle.internal.service.ServiceRegistry;
+import org.gradle.process.JavaExecSpec;
+import org.gradle.util.Path;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.File;
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+public class LifecycleAwareProject extends GroovyObjectSupport implements ProjectInternal, DynamicObjectAware {
+
+    public static ProjectInternal from(
+        ProjectInternal project,
+        IsolatedProjectEvaluationListenerProvider listenerProvider
+    ) {
+        return project instanceof LifecycleAwareProject
+            ? project
+            : new LifecycleAwareProject(project, listenerProvider);
+    }
+
+    private final ProjectInternal delegate;
+    private final IsolatedProjectEvaluationListenerProvider isolatedProjectEvaluationListenerProvider;
+    private final ExtensibleDynamicObject extensibleDynamicObject;
+
+    public LifecycleAwareProject(
+        ProjectInternal delegate,
+        IsolatedProjectEvaluationListenerProvider isolatedProjectEvaluationListenerProvider
+    ) {
+        this.delegate = delegate;
+        this.isolatedProjectEvaluationListenerProvider = isolatedProjectEvaluationListenerProvider;
+        this.extensibleDynamicObject = createExtensibleDynamicObject();
+    }
+
+    private ExtensibleDynamicObject createExtensibleDynamicObject() {
+        ExtensibleDynamicObject delegateExtensibleDynamicObject = (ExtensibleDynamicObject) ((DynamicObjectAware) delegate).getAsDynamicObject();
+        @SuppressWarnings("deprecation") ExtensibleDynamicObject extensibleDynamicObject = new ExtensibleDynamicObject(
+            this,
+            new BeanDynamicObject(this),
+            delegateExtensibleDynamicObject.getConvention()
+        );
+        extensibleDynamicObject.setParent(delegateExtensibleDynamicObject.getParent());
+        extensibleDynamicObject.addObject(delegateExtensibleDynamicObject, ExtensibleDynamicObject.Location.BeforeConvention);
+        return extensibleDynamicObject;
+    }
+
+    private void executeAllprojectsAction() {
+        isolatedProjectEvaluationListenerProvider.executeAllprojectsFor(delegate);
```

Comment:
🤔  This class does not need a full "evaluation listener provider" that's too broad of a dependency.
It is probably better to introduce a new lightweight service called something like `EagerLifecycleAllprojectsExecutor` that has only one method.

Reply:
Agree, in general